c/src/ml-api-remote-service.c

```diff
@@ -0,0 +1,389 @@
+
+  if (edge_info->topic)
+    nns_edge_set_info (edge_h, "TOPIC", edge_info->topic);
+
+  nns_edge_set_info (edge_h, "DEST_HOST", edge_info->dest_host);
+  sprintf (port, "%u", edge_info->dest_port);
+  nns_edge_set_info (edge_h, "DEST_PORT", port);
+}
+
+/**
+ * @brief Release edge info.
+ */
+static void
+_mlrs_release_edge_info (edge_info_s * edge_info)
+{
+  g_free (edge_info->dest_host);
+  g_free (edge_info->host);
+  g_free (edge_info->topic);
+}
+
+/**
+ * @brief Get ml remote service type from ml_option.
+ */
+static ml_remote_service_type_e
+_mlrs_get_service_type (gchar * service_str)
+{
+  ml_remote_service_type_e service_type = ML_REMOTE_SERVICE_TYPE_UNKNOWN;
+
+  if (g_ascii_strcasecmp (service_str, "model_raw") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_MODEL_RAW;
+  } else if (g_ascii_strcasecmp (service_str, "model_url") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_MODEL_URL;
+  } else if (g_ascii_strcasecmp (service_str, "pipeline_raw") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_PIPELINE_RAW;
+  } else if (g_ascii_strcasecmp (service_str, "pipeline_url") == 0) {
+    service_type = ML_REMOTE_SERVICE_TYPE_PIPELINE_URL;
+  } else {
+    _ml_error_report ("Invalid service type: %s, Please check service type.",
+        service_str);
+    service_type = ML_REMOTE_SERVICE_TYPE_UNKNOWN;
+  }
+  return service_type;
+}
+
+/**
+ * @brief Process ml remote service
+ */
+static void
+_mlrs_process_remote_service (nns_edge_data_h data_h)
+{
+  void *data;
+  nns_size_t data_len;
+  gchar *service_str = NULL;
+  gchar *service_key = NULL;
+  ml_remote_service_type_e service_type;
+
+  nns_edge_data_get (data_h, 0, &data, &data_len);
+
+  nns_edge_data_get_info (data_h, "service-type", &service_str);
+  service_type = _mlrs_get_service_type (service_str);
+  nns_edge_data_get_info (data_h, "service-key", &service_key);
+
+  switch (service_type) {
+    case ML_REMOTE_SERVICE_TYPE_MODEL_URL:
+      /** @todo Download the model file from given URL */
+    case ML_REMOTE_SERVICE_TYPE_MODEL_RAW:
+      /** @todo Save model file to given path and register the model */
+      break;
+    case ML_REMOTE_SERVICE_TYPE_PIPELINE_URL:
+      /** @todo Download the pipeline description from given URL */
+    case ML_REMOTE_SERVICE_TYPE_PIPELINE_RAW:
+      ml_service_set_pipeline (service_key, (gchar *) data);
+      break;
+    default:
+      _ml_error_report
+          ("Unknown service type or not supported yet. Service num: %d",
+          service_type);
+      break;
+  }
+}
+
+/**
+ * @brief Edge event callback.
+ */
+static int
+_mlrs_edge_event_cb (nns_edge_event_h event_h, void *user_data)
+{
+  nns_edge_event_e event = NNS_EDGE_EVENT_UNKNOWN;
+  nns_edge_data_h data_h = NULL;
+
+  int ret;
+
+  ret = nns_edge_event_get_type (event_h, &event);
+  if (NNS_EDGE_ERROR_NONE != ret)
+    return ret;
+
+  switch (event) {
+    case NNS_EDGE_EVENT_NEW_DATA_RECEIVED:{
+      nns_edge_event_parse_new_data (event_h, &data_h);
```

Comment:
I think It needs to check return value. because, ```nns_edge_data_copy ``` and ```nns_edge_metadata_copy``` in this function have some error case.

Reply:
Added to check return value. Thanks.